### ISSUE TITLE
Wrongly adding new currency for BNPLs

### ISSUE THREAD

**Issue by KarlisJ** [Automattic/woocommerce-payments#6861]
### Describe the bug
When adding any BNPL payment method (currently Affirm or Afterpay) multi-currency incorrectly tries to add new currencies to the store. 

Stripe handle allowed currencies for BNPLs differently from other PMs. For BNPLs currency has to be dynamically restricted to only domestic currency which was implemented for checkout in #6708.
### To Reproduce

- Have a US-based merchant account
- Enable multi-currency
- Remove CAD from multi-currency options (but keep multi-currency enabled)
- Try to add Affirm as LPM
- There should be a note indicating that CAD will be added to the store. 

### Actual behavior
We try to add CAD to the store. 

### Screenshots
![image](https://github.com/Automattic/woocommerce-payments/assets/24650518/0580c597-577c-42b0-b04b-1db2ff69e8f1)

### Expected behavior
We shouldn't add CAD to the store, because for US-based merchants only USD is available at checkout for Affirm.

The same problem occurs for CAN-based merchants for USD. Also for All possible cases for Afterpay.

### Additional context
The issue here is that we use a static array `client/payment-methods-map.tsx` to get currencies in JS. We should load it dynamically, from PHP similar to how we load the enabled payment method ids now. However, doing that requires a larger refactor because the code expects `paymentMethodsMap` to be populated on the first load. As soon as we populate it async, we're getting a lot of `Undefined` errors braking the whole setting screen.

**Comment by htdat** [comment]
This issue impacts `Multi-currency`, so assigning to Sigma (based on team responsibilities Pc2DNy-3z-p2) cc @dwainm as team lead. Assigning as part of Gamma Triage process PcreKM-yM-p2.

**Comment by KarlisJ** [comment]
Wanted to quickly note that we had discussion whether the approach we have now is correct p1689847026961299-slack-CGGCLBN58

TLDR: (there are 70+messages, most unrelated, best to start here p1690354533169209/1689847026.961299-slack-CGGCLBN58) 
We force merchants to add all supported currencies while for LPM to work they need only one.
However, there's uncertainty on how to handle case when merchants don't have any of supported currencies enabled. 

We'll create an RFC for this.

**Comment by jessepearson** [comment]
@KarlisJ Was an RFC ever created?

From what I can tell, the currencies are added via the code here:
https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/PaymentMethodsCompatibility.php

The `add_missing_currencies` method calls `get_enabled_payment_method_currencies` and searches for the payment method class to find its currency, like this: https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-affirm-payment-method.php#L32

I'm wondering if we can add an additional element to the array that is an array like this:

```
		$this->required_currencies          = [
			'any' => [ 'USD', 'CAD' ],
			'all' => [ 'USD', 'CAD' ],
		];
```

We would typically only need one or the other, but this would allow us to say that if USD is enabled, don't enable CAD and vice versa.

**Comment by KarlisJ** [comment]
@jessepearson No, RFC wasn't created. We were strongly time-boxed with the BNPL project and [implemented a solution ](https://github.com/Automattic/woocommerce-payments/pull/6708)using `$this->accept_only_domestic_payment = true;` 

Unless we want to somehow improve the mechanism, this issue can be closed.

**Comment by jessepearson** [comment]
@KarlisJ Okay, closing the issue based on your comment. Thank you.
